Marlin/Cartridge.cpp

```diff
@@ -0,0 +1,237 @@
+    if (IsSafetyCriticalSection()) {
+        static bool killed = false;
+        if (IsRunning()) {
+            SERIAL_ERROR_START;
+            serialprintPGM(serial_msg);
+            SERIAL_EOL;
+        }
+        if (!killed) {
+            Running = false;
+            killed = true;
+            kill(serial_msg);
+        }
+    } else {
+        static millis_t timeSinceLastRemoval = {0};
+        if (millis() > timeSinceLastRemoval + CARTRIDGE_REMOVED_ERR_INTERVAL) {
+            quickStop();
+            disable_all_heaters();
+            disable_all_steppers();
+            serialprintPGM(serial_msg);
+            SERIAL_EOL;
+            SERIAL_ECHOLN("// action:pause");
+        }
+        timeSinceLastRemoval = millis();
+    }
+}
+
+/**
+ * Macro function that updates cartridge status, and checks if a cartridge has 
+ * been removed.
+ * @returns    Returns true if a cartridge is removed
+ */
+bool CartridgeUpdateAndCheck() {
+    bool returnValue = false;
+    UpdateCartridgeStatus();
+    returnValue = CartridgeRemoved();
+    return returnValue;
+}
+
+//===========================================================================
+//============================ Private Functions ============================
+//===========================================================================
+
+/**
+ * Reports that a cartridge is absent. If there was a cartridge present,
+ * marks it as removed
+ */
+static void cartridgeAbsentUpdate(unsigned int cartNumber) {
+    if (cartridgePresent[cartNumber] == true) {
+        cartridgeRemoved[cartNumber] = true;
+        switch (cartNumber) {
+            case 0:
+                SERIAL_ECHOLN("Cartidge 0 Removed");
+                break;
+            case 1:
+                // Prevents the silver extruder from being lowered
+                // unintentionally
+                WRITE(CART1_SIG1_PIN, LOW);
+                SERIAL_ECHOLN("Cartidge 1 Removed");
+                break;
+            default:
+                SERIAL_ECHOLN("Cartridge Removed");
+        }
+    }
+    cartridgePresent[cartNumber] = false;
+}
+
+/**
+ * Reports that a cartridge is present. If it was marked as removed,
+ * this will clear it.
+ */
+static void cartridgePresentUpdate(unsigned int cartNumber) {
+    if (cartridgePresent[cartNumber] == false) {
+        switch (cartNumber) {
+            case 0:
+                SERIAL_ECHOLN("Cartidge 0 Inserted");
+                break;
+            case 1:
+                SERIAL_ECHOLN("Cartidge 1 Inserted");
+                break;
+            default:
+                SERIAL_ECHOLN("Cartridge Inserted");
+        }
+    }
+    cartridgePresent[cartNumber] = true;
+    cartridgeRemoved[cartNumber] = false;
+}
+
+/**
+ * Signals that no cartridges have been removed, used to tell if the system 
+ * is safe to run after a reset with cartridges removed.
+ * @returns    Returns true if no catridges have been removed, false otherwise.
```

Comment:
This looks like it returns true if cartridges _have_ been removed, unless I'm missing something?